attributecert/attributecert.go

```diff
@@ -0,0 +1,713 @@
+		rest, err := asn1.Unmarshal(rawAttr.FullBytes, &parsedAttribute)
+		if err == nil && len(rest) == 0 {
+			attributes = append(attributes, parsedAttribute)
+		} else {
+			return nil, err
+		}
+	}
+
+	for _, attribute := range attributes {
+		if attribute.Id.Equal(oidAttributeUserNotice) {
+			for _, value := range attribute.RawValues {
+				var userNotice userNotice
+				_, err := asn1.Unmarshal(value.FullBytes, &userNotice)
+				if err != nil {
+					return nil, err
+				}
+				out.UserNotice = userNotice
+			}
+		} else if attribute.Id.Equal(oidTcgPlatformSpecification) {
+			var platformSpecification TCGPlatformSpecification
+			_, err := asn1.Unmarshal(attribute.RawValues[0].FullBytes, &platformSpecification)
+			if err != nil {
+				return nil, err
+			}
+			out.TCGPlatformSpecification = platformSpecification
+		} else if attribute.Id.Equal(oidTbbSecurityAssertions) {
+			var securityAssertions TBBSecurityAssertions
+			_, err := asn1.Unmarshal(attribute.RawValues[0].FullBytes, &securityAssertions)
+			if err != nil {
+				return nil, err
+			}
+			out.TBBSecurityAssertions = securityAssertions
+		} else if attribute.Id.Equal(oidTcgCredentialSpecification) {
+			var credentialSpecification TCGCredentialSpecification
+			_, err := asn1.Unmarshal(attribute.RawValues[0].FullBytes, &credentialSpecification)
+			if err != nil {
+				return nil, err
+			}
+		} else if attribute.Id.Equal(oidTcgCredentialType) {
+			var credentialType TCGCredentialType
+			_, err := asn1.Unmarshal(attribute.RawValues[0].FullBytes, &credentialType)
+			if err != nil {
+				return nil, err
+			}
+		} else if attribute.Id.Equal(oidTcgPlatformConfigurationV1) {
+			var platformConfiguration PlatformConfigurationV1
+			_, err := asn1.Unmarshal(attribute.RawValues[0].FullBytes, &platformConfiguration)
+			if err != nil {
+				return nil, err
+			}
+		} else if attribute.Id.Equal(oidTcgPlatformConfigurationV2) {
+			var platformConfiguration PlatformConfigurationV2
+			_, err := asn1.Unmarshal(attribute.RawValues[0].FullBytes, &platformConfiguration)
+			if err != nil {
+				return nil, err
+			}
+		} else if attribute.Id.Equal(oidTcgPlatformConfigUri) {
+			var platformConfigurationUri UriReference
+			_, err := asn1.Unmarshal(attribute.RawValues[0].FullBytes, &platformConfigurationUri)
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			return nil, fmt.Errorf("unknown attribute %v", attribute.Id)
+		}
+	}
+
+	for _, extension := range in.TBSAttributeCertificate.Extensions {
+		if extension.Id.Equal(oidExtensionSubjectAltName) {
+			var seq asn1.RawValue
+			rest, err := asn1.Unmarshal(extension.Value, &seq)
+			if err != nil {
+				return nil, err
+			} else if len(rest) != 0 {
+				return nil, errors.New("x509: trailing data after X.509 extension")
+			}
+			rest = seq.Bytes
+			for len(rest) > 0 {
+				var v asn1.RawValue
+				rest, err = asn1.Unmarshal(rest, &v)
+				if err != nil {
+					return nil, err
+				}
+				if v.Tag == asn1.TagSet {
```

Comment:
The logic for pulling out the different fields seems like a good candidate to decompose into its own method, maybe an unexported `populateFromTCGData()` method?